Release v1.0.0

PhonikTop opened the pull request.

Str1kez commented on clocks/Dockerfile:
```diff
@@ -0,0 +1,24 @@
+COPY requirements.txt .
+RUN pip install --no-cache-dir -r requirements.txt
+
+COPY . .
```
Наверное на этом слое не нужен весь код, можно скопировать его в соответствующих

Str1kez commented on clocks/meetings/migrations/0001_initial.py:
все миграции можно в одну

Str1kez commented on clocks/meetings/models.py:
```diff
@@ -0,0 +1,16 @@
+
+
+class Meeting(models.Model):
+    room_id = models.ForeignKey(
```
```suggestion
    room = models.ForeignKey(
```
это же уже по сути объект

Str1kez commented on clocks/meetings/serializers.py:
```diff
@@ -0,0 +1,26 @@
+            allowed = set(fields)
+            existing = set(self.fields)
+            for field_name in existing - allowed:
+                self.fields.pop(field_name)
```
`self.fields` вроде list и так не удалишь

PhonikTop replied:
Я просто подумал, что так сериализатор всегда будет работать со всеми полями, определенными по умолчанию, игнорируя те параметры, что в него передают в других модулях проекта.

Str1kez replied:
да решение нормальное, сам метод pop у листа на индекс работает же

Str1kez replied:
и как будто можно было проще сделать без цикла 
```python
self.fields = list(existing - allowed)
```

Str1kez replied:
а там дикт в кэше, все ок тогда

Str1kez commented on clocks/meetings/serializers.py:
```diff
@@ -0,0 +1,26 @@
+
+
+class MeetingSerializer(serializers.ModelSerializer):
+    room_id = serializers.PrimaryKeyRelatedField(queryset=Room.objects.all())
+    task_name = serializers.CharField(max_length=100)
+    votes = serializers.JSONField()
+    average_score = serializers.FloatField(allow_null=True, required=False)
+    active = serializers.BooleanField()
```
разве это автоматом не делается `ModelSerializer`? Вроде для этого он и нужен

Str1kez commented on clocks/meetings/views.py:
```diff
@@ -0,0 +1,143 @@
+        room = get_object_or_404(Room, id=room_id)
+        serializer = MeetingSerializer(data=request.data, fields=["room_id", "task_name"])
+
+        if serializer.is_valid():
```
лучше наоборот not is_valid проверить и сразу вернуть ошибку, будет меньше вложенности для последующего кода

Str1kez replied:
следующее условие аналогично

Str1kez commented on clocks/meetings/views.py:
```diff
@@ -0,0 +1,143 @@
+
+    def get(self, request: Request, meeting_id: int) -> Response:
+        meeting = get_object_or_404(Meeting, id=meeting_id)
+        serializer = MeetingSerializer(meeting,
+                                       fields=["id", "room_id", "task_name", "votes", "average_score", "active"])
```
```suggestion
        serializer = MeetingSerializer(meeting)
```

Str1kez commented on clocks/api/api_utils.py:
```diff
@@ -0,0 +1,16 @@
+    def __init__(self):
+        # Создаем частичные функции с предустановленными параметрами
+        self.success_response = partial(self.return_api_response, type_message="success")
+        self.error_response = partial(self.return_api_response, type_message="error")
```
можно тут прокинуть сразу статус 400 и не передавать его в коде

PhonikTop replied:
> можно тут прокинуть сразу статус 400 и не передавать его в коде

А если нужно выводить другой статус ошибки?

Str1kez replied:
тогда будешь передавать его где надо, а по дефолту будет 400

Str1kez commented on clocks/meetings/views.py:
```diff
@@ -0,0 +1,143 @@
+                                           response_status=status.HTTP_400_BAD_REQUEST)
+
+
+class GetMeetingView(APIView):
```
Можно использовать [RetrieveApiView](https://www.django-rest-framework.org/api-guide/generic-views/#retrieveapiview)

Str1kez commented on clocks/meetings/views.py:
```diff
@@ -0,0 +1,143 @@
+    """
+
+    def post(self, request: Request, meeting_id: int) -> Response:
+        meeting = get_object_or_404(Meeting, id=meeting_id)
```
можно брать комнату и делать select_related митинга, так один запрос будет только
что -то по типу 
```python
room = get_object_or_404(Room.objects.select_related('current_meeting'), current_meeting_id=meeting_id)
```

Str1kez commented on clocks/meetings/views.py:
```diff
@@ -0,0 +1,143 @@
+    def post(self, request: Request, meeting_id: int) -> Response:
+        meeting = get_object_or_404(Meeting, id=meeting_id)
+
+        meeting.active = True
+        meeting.votes = {}
+        meeting.average_score = 0
+        meeting.save()
```
Можно в модели сделать метод `reset_to_default` и вызвать его тут

Str1kez commented on clocks/meetings/views.py:
```diff
@@ -0,0 +1,143 @@
+    """
+
+    def post(self, request: Request, meeting_id: int) -> Response:
+        meeting = get_object_or_404(Meeting, id=meeting_id)
```
аналогично с select_related

Str1kez commented on clocks/meetings/views.py:
```diff
@@ -0,0 +1,143 @@
+
+        serializer = MeetingSerializer(data=request.data, fields=["task_name"])
+
+        if serializer.is_valid():
```
проверять not is_valid

Str1kez commented on clocks/requirements.txt:
```diff
@@ -0,0 +1,8 @@
+Django~=5.1
```
лучше явно указывай полностью версию

Str1kez replied:
и версию с ==

Str1kez commented on clocks/requirements.txt:
```diff
@@ -0,0 +1,8 @@
+asgiref~=3.8.1
+channels~=4.1.0
+djangorestframework~=3.15.2
+redis~=5.1.0b7
```
не надо бета релизы юзать)

Str1kez commented on clocks/rooms/serializers.py:
```diff
@@ -0,0 +1,25 @@
+from meetings.models import Meeting
+from rest_framework import serializers
+
+from .models import Room
```
лучше использовать везде абсолютные пути импорта

PhonikTop replied:
> лучше использовать везде абсолютные пути импорта

У меня при запуске через runserver выходила ошибка 
```
from ..meetings.models import Meeting
ImportError: attempted relative import beyond top-level package
```

Str1kez replied:
Это же опять относительные. Попробуй `clocks.meetings.models import Meeting`

Str1kez commented on clocks/rooms/views.py:
```diff
@@ -0,0 +1,104 @@
+response = APIResponseHandler()
+
+
+class RoomCreateView(APIView):
```
CreateAPIView, можно больше использовать дженериков -> меньше кода

Str1kez commented on clocks/ws/consumers.py:
почему решил не делать async консюмера?

PhonikTop replied:
> почему решил не делать async консюмера?

Я начал, но у меня из-за этого была ошибка связанная с ретурном данных о голосах. Я возился с ней больше дня, но ничего не помогало. Решил просто не делать это, оставил код синхронным, но рабочим.

Str1kez replied:
а в чем именно ошибка была? Запросы к бд обернул в `sync_to_async`?

PhonikTop replied:
> а в чем именно ошибка была? Запросы к бд обернул в `sync_to_async`?

Да. Ошибка была в том, что объект корутины не может возвращаться. Я подумал, что это из-за вывода в самом декораторе, решил, что пока не смогу с этим справиться.

Str1kez commented on clocks/requirements.txt:
```diff
@@ -0,0 +1,8 @@
+redis~=5.1.0b7
+django-grappelli~=4.0.1
+uvicorn
+daphne
```
наверное можно дропнуть daphne

PhonikTop replied:
> наверное можно дропнуть daphne

На нем же web socket держится

Str1kez replied:
не может быть такого. У тебя же uvicorn есть

Str1kez commented on clocks/rooms/models.py:
```diff
@@ -0,0 +1,17 @@
+class Room(models.Model):
+    name = models.CharField(max_length=100)
+    is_active = models.BooleanField(default=True)
+    current_meeting = models.OneToOneField(
```
Вроде и не нужно, по сути один митинг может быть активен для комнаты

PhonikTop replied:
> Вроде и не нужно, по сути один митинг может быть активен для комнаты

В каком смысле?

Str1kez replied:
Непонятно зачем нужен current_meeting, ведь текущий митинг можно найти `Room.objects.filter(meetings__is_active=True)`

PhonikTop replied:
Подумал, что так будет читабельнее. Плюс, на момент, когда писал это, не догадывался о существовании фильтра.

Str1kez replied:
тогда оставь тут, а в meeting удали FK и поправь везде код

Str1kez commented on Dockerfile:
```diff
@@ -0,0 +1,42 @@
+
+COPY . .
+
+RUN tree
+
+WORKDIR /Clocks/clocks
+RUN python manage.py migrate
+RUN python manage.py collectstatic --noinput
```
не надо ничего мигрировать на этапе сборки образа

Str1kez commented on Dockerfile:
```diff
@@ -0,0 +1,42 @@
+FROM py-builder AS watchy-api-local
+
+CMD service nginx start && \
+    daphne -b 0.0.0.0 -p 8001 settings.asgi:application & \
```
для чего нужен?

PhonikTop replied:
Что именно? Daphne? Nginx?

Str1kez replied:
daphne

Str1kez commented on nginx/default.conf:
```diff
@@ -0,0 +1,28 @@
+        proxy_set_header X-Forwarded-Proto $scheme;
+    }
+
+    location /ws/ {
+        proxy_pass http://watchy-api:8001;
+        proxy_set_header Host $host;
+        proxy_set_header X-Real-IP $remote_addr;
+        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
+        proxy_set_header X-Forwarded-Proto $scheme;
+
+        proxy_http_version 1.1;
+        proxy_set_header Upgrade $http_upgrade;
+        proxy_set_header Connection "Upgrade";
+    }
```
можно вместе в daphne дропнуть как будто

Str1kez commented on clocks/meetings/views.py:
```diff
@@ -0,0 +1,150 @@
+        room_id = self.request.data.get("room")
+        task_name = self.request.data.get("task_name")
+
+        room = get_object_or_404(Room, id=room_id)
```
сериализатор уже достал объект, нет?

PhonikTop replied:
> сериализатор уже достал объект, нет?

Почему? это же уже другая модель. Я думал, он так не может

Str1kez replied:
а, да, ты прав

Str1kez commented on clocks/meetings/views.py:
```diff
@@ -0,0 +1,150 @@
+        room.save()
+        return meeting
+
+    def create(self, request, *args, **kwargs):
+        response = super().create(request, *args, **kwargs)
+        return Response({
+            "msg": "Meeting started",
+            "data": response.data
+        }, status=status.HTTP_201_CREATED)
```
лишнее

Str1kez commented on clocks/meetings/views.py:
```diff
@@ -0,0 +1,150 @@
+    serializer_class = MeetingSerializer
+    lookup_field = "id"
+
+    def retrieve(self, request: Request, *args, **kwargs) -> Response:
+        meeting = self.get_object()
+        serializer = self.get_serializer(meeting,
+                                         fields=["id", "room", "task_name", "votes", "average_score", "active"])
+        return response.success_response(msg="Meeting info", data=serializer.data, response_status=status.HTTP_200_OK)
```
Разве это не логика работы RetrieveAPIView?

Str1kez commented on clocks/meetings/views.py:
```diff
@@ -0,0 +1,150 @@
+    """
+    queryset = Meeting.objects.all()
+    serializer_class = MeetingSerializer
+    lookup_field = "id"
```
лишнее

PhonikTop replied:
А как он будет получать эти переменные? У GenericAPIView lookup_field = 'pk', а queryset serializer_class оба None

Str1kez replied:
так pk у тебя и есть id

Str1kez commented on clocks/rooms/views.py:
```diff
@@ -0,0 +1,104 @@
+    permission_classes = [IsAuthenticated]
+    serializer_class = RoomSerializer
+
+    def create(self, request: Request, *args, **kwargs) -> Response:
+        serializer = self.get_serializer(data=request.data, fields=["id", "name"])
+        if serializer.is_valid():
+            room = serializer.save()
+            response_serializer = self.get_serializer(instance=room, fields=["id", "name"])
+            return response.success_response(msg="Room created", data=response_serializer.data,
+                                             response_status=status.HTTP_201_CREATED)
+        return response.error_response(msg="Error", data=serializer.errors,
+                                       response_status=status.HTTP_400_BAD_REQUEST)
```
тоже лишнее, можно все данные отдавать при создании

Str1kez commented on clocks/rooms/views.py:
```diff
@@ -0,0 +1,104 @@
+    serializer_class = RoomSerializer
+    lookup_field = "id"
+
+    def get_queryset(self):
+        return Room.objects.all()
```
`queryset = `

Str1kez commented on clocks/rooms/views.py:
```diff
@@ -0,0 +1,104 @@
+        return Room.objects.all()
+
+
+class RoomDetailView(RetrieveAPIView, DestroyAPIView):
```
```suggestion
class RoomDetailView(RetrieveDestroyAPIView):
```

Str1kez commented on clocks/rooms/views.py:
```diff
@@ -0,0 +1,104 @@
+            return [IsAuthenticated()]
+        return [AllowAny()]
+
+    def delete(self, request: Request, *args, **kwargs) -> Response:
+        room = self.get_object()
+        room.delete()
+        return response.success_response(msg="Room deleted successfully", data=None,
+                                         response_status=status.HTTP_204_NO_CONTENT)
```
```suggestion
```

Str1kez commented on clocks/rooms/views.py:
```diff
@@ -0,0 +1,104 @@
+        return response.success_response(msg="Room participants", data=data,
+                                         response_status=status.HTTP_200_OK)
+
+    def get_object(self):
+        return Room.objects.get(id=self.kwargs.get("id"))
```
`queryset=`

PhonikTop replied:
Она же тут должна получать айди комнаты.

Str1kez replied:
ну да, в url передаешь и логика подефолту так работает, если `queryset` указан. Если не хочешь через `queryset`, то хотя бы `get_object_or_404` используй

PhonikTop replied:
Так через get_object_or_404 выдает только один объект, а мне нужно несколько. А если через queryset, то он не фильтрует по значению из lookup_fields и выдает просто все активные комнаты.

Str1kez commented on clocks/rooms/views.py:
```diff
@@ -0,0 +1,104 @@
+        data = [
+            {
+                "token": token,
+                "role": "observer" if role == "observer" else "voter",
```
роли в Enum лучше

Str1kez replied:
и зачем это все, если можно просто передавать `room.users`?

Str1kez replied:
вообще лучше выпилить это и добавить что-то такое
```python
    def get_serializer(self, *args, **kwargs):
        kwargs['fields'] = ['users']
        return super().get_serializer(*args, **kwargs)
```

Str1kez replied:
UPD: Аналогично, напиши сериалайзер для получения данных о комнате, которых будет возвращать такой dict

Str1kez commented on clocks/rooms/views.py:
```diff
@@ -0,0 +1,104 @@
+    serializer_class = MeetingSerializer
+    lookup_field = "id"
+
+    def get_queryset(self):
+        room = Room.objects.get(id=self.kwargs.get("id"))
+        return Meeting.objects.filter(room=room, active=False)
```
`queryset = ` просто поменяй lookup_field

Str1kez commented on clocks/settings/asgi.py:
```diff
@@ -1,16 +1,19 @@
+django_asgi_app = get_asgi_application()
+
+
+from ws.rounting import ws_urlpatterns
```
```suggestion
from ws.routing import ws_urlpatterns
```

Str1kez commented on requirements.txt:
```diff
@@ -0,0 +1,9 @@
+Django~=5.1.1
```
```suggestion
Django==5.1.1
```

Str1kez commented on Dockerfile:
```diff
@@ -0,0 +1,44 @@
+
+FROM py-builder AS watchy-api-prod
+
+RUN python manage.py migrate
+RUN python manage.py collectstatic --noinput
```
мало того, что не убрал, так еще и тут добавил  😂

PhonikTop replied:
> мало того, что не убрал, так еще и тут добавил 😂

а куда их деть тогда🤕

Str1kez replied:
отдельный скрипт в папку `bin/migrations.sh`

Str1kez commented on clocks/meetings/serializers.py:
```diff
@@ -0,0 +1,19 @@
+class MeetingSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = Meeting
+        fields = ["id", "room", "task_name", "votes", "average_score", "active"]
```
```suggestion
        fields = '__all__'
```

Str1kez replied:
```suggestion
        fields = '__all__'
```

Str1kez commented on clocks/meetings/views.py:
```diff
@@ -0,0 +1,104 @@
+    queryset = Meeting.objects.all()
+    serializer_class = MeetingSerializer
+
+    def get_serializer(self, *args, **kwargs):
```
зачем? там ведь по дефолту все поля

Str1kez commented on clocks/meetings/views.py:
```diff
@@ -0,0 +1,104 @@
+
+
+class EndMeetingView(UpdateAPIView):
+    queryset = Meeting.objects.all()
```
```suggestion
    queryset = Meeting.objects.select_related('room').filter(active=True)
```

Str1kez replied:
не будет доп запроса для комнаты и сразу активный будет искаться

Str1kez commented on clocks/meetings/views.py:
```diff
@@ -0,0 +1,104 @@
+    queryset = Meeting.objects.all()
+    serializer_class = MeetingSerializer
+
+    def update(self, request, *args, **kwargs):
```
нужно переопределять `perform_update`, а то здесь весь функционал `UpdateAPIView` теряется

Str1kez commented on clocks/meetings/views.py:
```diff
@@ -0,0 +1,104 @@
+    queryset = Meeting.objects.all()
+    serializer_class = MeetingSerializer
+
+    def update(self, request, *args, **kwargs):
```
аналогично не тот метод

Str1kez commented on clocks/meetings/views.py:
```diff
@@ -0,0 +1,104 @@
+        return response.success_response(msg="Meeting Restarted", response_status=status.HTTP_200_OK)
+
+
+class UpdateMeetingTaskView(UpdateAPIView):
+    queryset = Meeting.objects.all()
+    serializer_class = MeetingSerializer
+
+    def update(self, request, *args, **kwargs):
+        serializer = self.get_serializer(self.get_object(), data={"task_name": request.data.get("task_name")},
+                                         partial=True)
+
+        if not serializer.is_valid():
+            return response.error_response(msg="Error", data=serializer.errors,
+                                           response_status=status.HTTP_400_BAD_REQUEST)
+
+        serializer.save()
+        return response.success_response(msg="Task updated", data={"task_name": request.data.get("task_name")},
+                                         response_status=status.HTTP_200_OK)
```
Подход неправильный, нужен отдельный сериалайзер для обновления митинга, тогда не надо будет костылить так вьюшки. И вообще надо разделять сериализаторы, один для чтения, другой для создания, другой для обновления. А то ты один сериализатор мучаешь с динамическими `fields`

Str1kez replied:
RoomSerializer в эту же тему

Str1kez commented on clocks/meetings/views.py:
```diff
@@ -0,0 +1,104 @@
+
+    def retrieve(self, request, *args, **kwargs):
+        meeting = self.get_object()
+        meeting.average_score = round(
+            sum(map(int, meeting.votes.values())) / len(meeting.votes)) if meeting.votes else 0
+        meeting.save()
```
это не `POST` запрос, не должно быть тут изменения модели

PhonikTop replied:
> это не `POST` запрос, не должно быть тут изменения модели

а если поменяю на UpdateAPIView?

Str1kez commented on clocks/rooms/serializers.py:
```diff
@@ -0,0 +1,19 @@
+class RoomSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = Room
+        fields = ["id", "name", "is_active", "users", "current_meeting_id"]
```
```suggestion
        fields = '__all__'
```

Str1kez commented on clocks/rooms/views.py:
```diff
@@ -0,0 +1,87 @@
+    """
+    serializer_class = MeetingSerializer
+
+    def get_queryset(self):
+        return Meeting.objects.filter(room_id=self.kwargs.get("pk"))
```
Тоже не нужно, задай 
```
queryset = Meeting.objects.filter(active=False)
lookup_field = 'room_id'
```
сюда же можешь добавить пагинатор

PhonikTop replied:
А как он будет фильтровать queryset по нужной комнате?

Str1kez replied:
он же в методе get_object по lookup_field вызывает get_or_404

Str1kez commented on requirements.txt:
```diff
@@ -0,0 +1,9 @@
+Django==5.1.1
+asgiref~=3.8.1
```
так же как с джангой, укажи везде точные версии. Некоторые пакеты вообще без версий

PhonikTop committed:
Переписал JoinRoomView под GenericAPIView и сделал создание куки для новых клиентов

Str1kez commented on clocks/meetings/models.py:
```diff
@@ -0,0 +1,35 @@
+        self.average_score = None
+        self.save()
+
+    def end_meeting(self):
+        self.active = False
+        self.room.current_meeting = None
+        self.room.users = []
+        self.room.save()
+        self.save()
+
+    def get_average_score(self):
+        from .logic import get_average_score
+        self.average_score = get_average_score(self.votes)
+        self.save()
```
тогда это все уж в logic

Str1kez replied:
и импорт каждый раз = плохо

Str1kez replied:
и почему get_, ты тут ничего не возвращаешь

PhonikTop replied:
> тогда это все уж в logic

А почему?

PhonikTop replied:
> тогда это все уж в logic

Там же нету никакой логики, чтобы выносить её туда. Только изменение данных о конкретном голосовании и очистка списка пользователей связанной комнаты.

Str1kez replied:
все что ты написал и есть бизнес логика)) А тут получается модель использует бизнес логику, чего быть не должно

Str1kez commented on clocks/meetings/models.py:
```diff
@@ -0,0 +1,35 @@
+    active: bool = models.BooleanField(default=True)
+
+    def __str__(self) -> str:
+        return f"Meeting for {self.room.name} - {self.task_name}"
```
зачем дергать второй раз ради получения строки? Поменяй на self.room_id

Str1kez commented on clocks/users/views.py:
```diff
@@ -0,0 +1,52 @@
+
+        redis.save_new_client_to_redis(cookie, nickname, role)
+
+        room = get_object_or_404(Room, id=self.request.data.get("room_id"))
```
слишком много лишних запросов в бд в этом методе, и почему проверка существования комнаты почти в конце? Если ее нет, зачем все выше обрабатывать? И наверное лучше комнату как path параметр в url

PhonikTop replied:
> И наверное лучше комнату как path параметр в url

почему?

Str1kez replied:
Потому что по сути это `UpdateApiView` для `Room`. Особенно первые строки метода это и есть получение объекта по кверисету и `id` по `lookup_field`

PhonikTop committed:
Добавил переменные окружения

PhonikTop committed:
Сделал новый сериализатор для списка голосов

PhonikTop committed:
Добавил асинхронный BaseConsumer и почистил код

PhonikTop committed:
Сделал проверку на существование активного голосования

PhonikTop committed:
Убрал не используемый метод chat_message

PhonikTop committed:
Сделал завершение голосования после того, как проголосовал последний игрок

PhonikTop committed:
Сделал сохранение зашифрованный token id в куки пользователя

PhonikTop committed:
Добавил отправку сообщения в группу WebSocket при подключении нового участника

PhonikTop committed:
Убрал запуск нового голосования при входе пользователя в комнату без активного meeting

PhonikTop committed:
Реализована динамическая настройка путей для поиска приложений

PhonikTop committed:
Разделены `INSTALLED_APPS` на `DJANGO_APPS`, `THIRD_APPS` и `LOCAL_APPS`

Str1kez commented on clocks/meetings/views.py:
```diff
@@ -0,0 +1,77 @@
+
+        meeting = serializer.save(room=room, task_name=self.request.data.get("task_name"))
+        room.current_meeting = meeting
+        room.save()
```
комната же в сериализаторе достается + валидацию можно в нем же сделать

Str1kez commented on clocks/meetings/views.py:
```diff
@@ -0,0 +1,77 @@
+
+        if not meeting.active:
+            raise ValidationError({"error": "Meeting already completed."})
+
```
не будет такого по кверисету

Str1kez commented on clocks/rooms/views.py:
```diff
@@ -0,0 +1,64 @@
+    """
+    serializer_class = MeetingHistorySerializer
+
+    def get_queryset(self):
+        return get_list_or_404(Meeting, room_id=self.kwargs.get("pk"))
```
почему не просто через кверисет в классе?

PhonikTop replied:
Я не нашел, как это сделать для списка

Str1kez replied:
Ты не нашел как сделать так, чтобы при пустом списке отдавало 404? Если да, то оно тебе надо?) Можно отдавать 200 с пустым списком, это ок, ведь ты запрашиваешь коллекцию объектов, а не единичный

Str1kez replied:
К тому же переопределять выдачу `queryset`, который можно разбить на страницы (чтобы выгружать меньше данных), на список всех сессий - не очень хорошо

PhonikTop replied:
> Ты не нашел как сделать так, чтобы при пустом списке отдавало 404?


Нееее. Почему-то с queryset выдаются даже те голосования, которые не привязаны к конкретно запрашиваемой комнате.

Str1kez replied:
не может быть такого, если ты правильно указал queryset
```
queryset = Meeting.objects.all()
lookup_field = 'room_id'
```

Str1kez replied:
С этим как дела обстоят?

Str1kez commented on clocks/settings/settings.py:
```diff
@@ -1,45 +1,55 @@
-DEBUG = True
 
-ALLOWED_HOSTS = []
+sys.path.insert(1, PROJECT_ROOT)
+for app_lookup_path in ("clocks",):
+    sys.path.insert(1, at_project_root(app_lookup_path))
 
```
странности какие-то

PhonikTop replied:
> странности какие-то

почему?

PhonikTop replied:
> странности какие-то
> 

Я пытался сделать абсолютные импорты. Серёга посоветовал добавить этот код в settings, но у меня ничего не вышло. В итоге он сказал, что можно оставить импорты в том виде, в каком они у меня уже есть. Я решил сохранить этот код на всякий случай – он не мешает работе, а если в будущем разберусь, как исправить ошибку с абсолютными импортами, мне не придётся заново это писать.

Str1kez commented on clocks/users/redis_client.py:
```diff
@@ -0,0 +1,32 @@
+)
+
+
+class RedisClient:
```
А это зачем все? Почему не использовать кэш джанги через редис?

Str1kez commented on clocks/users/views.py:
```diff
@@ -0,0 +1,77 @@
+    serializer_class = UserInputSerializer
+
+    def get_object(self):
+        token = cookie_utils.cookie_decrypt(str(self.request.COOKIES.get("user")))
```
~exceptions не обрабатываешь~, оказывается ты внутри обрабатываешь, хотя мог сюда свои Exception от данного сервиса прокидывать, чтобы явно понимать, что шифрование пошло не так

Str1kez commented on clocks/api/api_utils.py:
```diff
@@ -0,0 +1,30 @@
+from settings.settings import CRYPT_KEY
+
+
+class Cookies_utils:
```
а в чем смысл класса? Можно отдельными функциями, а Fernet в глобальную

PhonikTop committed:
Вынес python manage.py migrate и python manage.py collectstatic из Dockerfile в api_entrypoint.sh

PhonikTop committed:
Сделал базу данных PostgreSQL

PhonikTop committed:
Исправил docker compose

Str1kez reviewed (commented):
оказывается вот как это работает, лоол

PhonikTop committed:
Исправил версии зависимостей

PhonikTop committed:
Сделал использование кэша django cache через redis

PhonikTop committed:
Сделал шифрование и расшифрование отдельными функциями.

PhonikTop committed:
Добавил документацию для инструментов API

PhonikTop committed:
Небольшие правки в settings.py

PhonikTop committed:
Исправил ошибку

PhonikTop committed:
Оптимизация взаимодействия с базой данных в JoinRoomView

PhonikTop committed:
Сделал валидацию в сериализаторе

PhonikTop committed:
Убрал лишний код

PhonikTop committed:
Исправил импорт

PhonikTop committed:
Перенес все конфигурационные файлы в отдельные директории

PhonikTop committed:
Создал директорию scripts

PhonikTop committed:
Перенес templates и static в директорию проекта

PhonikTop committed:
Исправил ошибку в web socket

Str1kez commented on clocks/meetings/views.py:
```diff
@@ -0,0 +1,67 @@
+
+class EndMeetingView(UpdateAPIView):
+    queryset = Meeting.objects.select_related("room").filter(active=True)
+    serializer_class = MeetingRemoveSerializer
+
+    def perform_update(self, request, *args, **kwargs):
+        meeting = self.get_object()
+        meeting.end_meeting()
```
1. Как будто тут и не нужен сериализатор, а хватит `lookup_field`
2. В `perform_update` другие аргументы прилетают, а у тебя сигнатура `update`

Str1kez replied:
в `RestartMeetingView` аналогично

Str1kez commented on clocks/meetings/views.py:
```diff
@@ -0,0 +1,67 @@
+
+class MeetingResultsView(UpdateAPIView):
+    queryset = Meeting.objects.all()
+    serializer_class = MeetingResultsSerializer
+
+    def perform_update(self, request, *args, **kwargs):
+        meeting = self.get_object()
+        meeting.get_average_score()
+        meeting.save()
```
1. В этот сериализатор как будто можно в запросе передать `average_score`, а это поле не должно ли быть `read-only`?
2. И как будто `perform_update` работает после инициализации сериализатора и в результат не попадет нововычисленный `average_score`
3. Это так же сигнатура `update`, так как в `perform_update` передается только сериализатор

PhonikTop replied:
> 1. В этот сериализатор как будто можно в запросе передать `average_score`, а это поле не должно ли быть `read-only`?

Немного не понял.

Str1kez replied:
Да, там кусок кода другой, но все то же. В ручку можно передать average_score, ведь сериализатор это позволяет. Второй пункт тоже остается нерабочим

Str1kez commented on clocks/settings/settings.py:
```diff
@@ -69,23 +81,11 @@
-# Password validation
-# https://docs.djangoproject.com/en/5.1/ref/settings/#auth-password-validators
+DATABASES = {}
+DATABASES["default"] = db = dj_database_url.parse(
+    f"postgres://{get_env_param_str('POSTGRES_USER')}:{get_env_param_str('POSTGRES_PASSWORD')}@watchy_db:5432/{get_env_param_str('POSTGRES_DB')}")
```
А почему не сразу закинуть в env переменную?
```env
DATABASE_URL=postgres://user:password@watchy_db:5432/clocks
```

PhonikTop replied:
> А почему не сразу закинуть в env переменную?
> 
> ```dotenv
> DATABASE_URL=postgres://user:password@watchy_db:5432/clocks
> ```

Эти переменные используются еще и для бд. 
Это удобно тем, что не нужно переписывать эти переменные вручную для апи из бд. Для смены пароля/логина, нужно просто изменить одну переменную в .env файле, а не переписывать дополнительно еще другую переменные в docker-compose.yml

Str1kez replied:
а кому надо будет менять пароль и юзера для локальной бд? 😂

Str1kez commented on clocks/users/redis_client.py:
```diff
@@ -0,0 +1,22 @@
+from django.core.cache import cache
+
+
+def save_new_client_to_cache(token, nickname: str, role: str, ttl: int = 432000) -> None:
```
```suggestion

DEFAULT_TOKEN_TTL = 60 * 60 * 24 * 5  # 5 days

def save_new_client_to_cache(token, nickname: str, role: str, ttl: int = DEFAULT_TOKEN_TTL) -> None:
```

так гораздо понятнее будет

Str1kez commented on clocks/users/redis_client.py:
```diff
@@ -0,0 +1,22 @@
+    """
+    Проверка на существование записи в кэше.
+    """
+    return cache.get(token) is not None
```
можно проще без взятия объекта. Такое решение работает, только с ttl у объекта, как раз твой случай
```suggestion
    return bool(cache.ttl(token))
```

Str1kez commented on clocks/users/views.py:
```diff
@@ -0,0 +1,78 @@
+        if cookie is None:
+            response.set_cookie("user", value=cookie_encrypt(token), max_age=432000)
+
+        if check_token_in_cache(token):
+            raise ValidationError({"error": "User exists"})
```
Проверка токена должна быть в самом начале, а то много логики уже сделал, а проверил только тут

PhonikTop replied:
> Проверка токена должна быть в самом начале, а то много логики уже сделал, а проверил только тут

Так все, что сверху, тоже часть проверки.

Str1kez replied:
Нет, ты просто все здесь запутал) У тебя токена зависит только от куки, но лишнюю работу по проверке данных ты уже сделал, хотя куку проверять можно и в начале

Str1kez commented on clocks/users/views.py:
```diff
@@ -0,0 +1,78 @@
+    serializer_class = UserInputSerializer
+
+    def get_object(self):
+        token = cookie_decrypt(str(self.request.COOKIES.get("user")))
```
сначала лучше проверь, что кука вообще есть, а потом возможный токен можно отправить на расшифровку

Str1kez replied:
Потому что расшифровать `'None'` - лишняя работа

Str1kez commented on clocks/users/views.py:
```diff
@@ -0,0 +1,78 @@
+        if token is None:
+            raise ValidationError("User cookie not valid")
+        user_data = get_client_data_by_token(token)
+        if not all(user_data.values()):
```
```suggestion
        if not user_data:
```

Str1kez commented on clocks/ws/consumers.py:
```diff
@@ -0,0 +1,101 @@
+            text_data_json = json.loads(text_data)
+            action = text_data_json.get("action")
+            if action:
+                response = await getattr(self, action)(text_data_json)
```
`action` не тот, то надо обработать исключение

Str1kez commented on clocks/ws/consumers.py:
```diff
@@ -0,0 +1,101 @@
+        return self.scope["url_route"]["kwargs"]["id"]
+
+    async def get_object(self, model, **filters):
+        return await database_sync_to_async(get_object_or_404)(model, **filters)
```
```suggestion
        return await aget_object_or_404(model, **filters)
```

Str1kez commented on clocks/ws/consumers.py:
```diff
@@ -0,0 +1,101 @@
+
+    async def get_lookup_id(self):
+        scope = self.scope["url_route"]["kwargs"]["id"]
+        exists = await database_sync_to_async(
+            lambda: Meeting.objects.filter(room_id=scope, active=True).exists()
+        )()
```
```suggestion
        exists = await Meeting.objects.filter(room_id=scope, active=True).aexists()
```

Str1kez commented on clocks/ws/consumers.py:
Кстати, не рассматривал этот пакет? Он как будто сильно упростил бы жизнь

https://github.com/NilCoalescing/djangochannelsrestframework

PhonikTop replied:
> Кстати, не рассматривал этот пакет? Он как будто сильно упростил бы жизнь
> 
> https://github.com/NilCoalescing/djangochannelsrestframework

Выглядит интересно, но как-то сомнительно. Не будет ли с ней еще больше проблем, и сможет ли кто-то другой разобраться в моем коде.🤷🏻‍♂️

Str1kez replied:
он вроде как раз делает работу с вебсокетом похожей на обычные generic views

PhonikTop replied:
> Кстати, не рассматривал этот пакет? Он как будто сильно упростил бы жизнь
> 
> https://github.com/NilCoalescing/djangochannelsrestframework

Я попробовал ее использовать, но она не предназначена для работы с группами и методов. В ней нету методов для отправки сообщений в группу или для создания/удаления самой группы.

Str1kez replied:
вообще-то есть, ну ок)

Str1kez commented on docker-compose.prod.yml:
```diff
@@ -0,0 +1,65 @@
+      - watchy_redis
+      - watchy_db
+    volumes:
+      - static:/Clocks/static
```
а зачем две статики? Вроде на nginx хватает

Str1kez commented on nginx/prod.conf:
вот это вряд ли так будет... и статики нет

PhonikTop committed:
Переместил конфигурации обратно в главную директорию

PhonikTop committed:
Заменил скрипт api_entrypoint.sh на Makefile

PhonikTop committed:
Merge pull request #4

Рефакторинг файловой структуры проекта

PhonikTop committed:
Исправил код

Co-authored-by: Ilya <68309482+Str1kez@users.noreply.github.com>

PhonikTop committed:
Убрал неиспользуемый prod.conf для nginx

PhonikTop committed:
Исправил код

PhonikTop committed:
Исправил код

PhonikTop committed:
Исправил работу EndMeetingView, RestartMeetingView и MeetingResultsView

PhonikTop committed:
Исправил работу EndMeetingView, RestartMeetingView и MeetingResultsView

PhonikTop committed:
Заменил get_list_or_404 на queryset в ручке RoomHistoryView

PhonikTop committed:
Перенес методы end_meeting и meeting_results в logic.py

PhonikTop committed:
Исправил нерабочий код

PhonikTop committed:
Добавил использование queryset в JoinRoomView

PhonikTop committed:
Изменил хранение информации о пользователях в голосовании

Str1kez commented on clocks/users/views.py:
```diff
@@ -0,0 +1,78 @@
+
+        nickname, role = serializer.validated_data["nickname"], serializer.validated_data["role"]
+        cookie = self.request.COOKIES.get("user")
+        token = cookie_decrypt(cookie) if cookie else str(uuid.uuid4())
```
лучше рейзить исключение, иначе сохранишь token = None

Str1kez commented on clocks/users/views.py:
```diff
@@ -0,0 +1,78 @@
+    serializer_class = UserInputSerializer
+
+    def get_object(self):
+        cookie = str(self.request.COOKIES.get("user") or ValidationError("User cookie is empty"))
+        token = cookie_decrypt(cookie)
+        if token is None:
+            raise ValidationError("User cookie is not valid")
```
это вообще не относится к логике получения объекта, надо вынести выше на слой проверки аутентификации

Str1kez commented:
И еще вопрос по моделям, почему бы не добавить created, updated?

PhonikTop commented:
> И еще вопрос по моделям, почему бы не добавить created, updated?

А для чего они?

Str1kez commented:
> > И еще вопрос по моделям, почему бы не добавить created, updated?
> 
> А для чего они?

чтобы видеть, когда была создана комната, таска

PhonikTop committed:
Сделал подведение результатов голосования после того, как все участники проголосовали.

PhonikTop committed:
Убрал переменные окружения из подключения к бд

PhonikTop committed:
Добавил поля created, updated для Room и Meeting

PhonikTop committed:
Улучшение обработки отсутствующих cookie

PhonikTop committed:
Исправил пути в docker-compose.yml

PhonikTop committed:
Добавил в админ-панель отображение истории комнаты

PhonikTop committed:
Исправил путь STATIC_ROOT в settings.py

PhonikTop committed:
Сделал вывод данных читабельнее

PhonikTop committed:
Добавил пагинацию для вывода истории комнаты

PhonikTop committed:
Доработал Makefile

PhonikTop committed:
Исправил валидацию ввода голоса пользователя

PhonikTop committed:
Обновил Makefile

PhonikTop committed:
Убрал лишнее

PhonikTop committed:
Исправил ошибку

PhonikTop committed:
Сделал кеширование данных голосов

PhonikTop committed:
Убрал current_meeting из модели Room

PhonikTop committed:
Сделал кеширование данных участников комнаты

PhonikTop committed:
Исправил ошибку

PhonikTop committed:
Создал класс для работы с данными голосований в кеше

PhonikTop committed:
Исправил Makefile

PhonikTop committed:
Улучшил структуру кода

PhonikTop committed:
Сделал корректный вывод данных голосов участников в истории комнаты

PhonikTop committed:
Переход на аутентификацию через Django Session

PhonikTop committed:
Переход на аутентификацию через Django Session

PhonikTop committed:
Создал функцию для отправки сообщения о присоединении нового участника в группу каналов по идентификатору комнаты

PhonikTop committed:
Перенес проверку на существование пользователя в redis_client.py

PhonikTop committed:
Перенос проверки активного голосования в сериализатор

PhonikTop committed:
Убрал не используемые импорты

PhonikTop committed:
Исправил возможность передачи average_score в запросе

PhonikTop committed:
Переделал веб-сокеты

PhonikTop committed:
Улучшение логики и оптимизация кода

Str1kez commented on docker-compose.yml:
```diff
@@ -0,0 +1,61 @@
+      - watchy_network
+
+  watchy-db:
+    image: postgres:14-alpine
```
17 уже вышел)

Str1kez commented on Makefile:
```diff
@@ -0,0 +1,52 @@
+status:
+	@docker-compose ps
+
+restart: stop start
```
`docker-compose restart` не подходит?

Str1kez commented on Makefile:
```diff
@@ -0,0 +1,52 @@
+status:
+	@docker-compose ps
+
+restart: stop start
```
`docker-compose restart` не подходит?

PhonikTop replied:
тогда будет бд сбрасывать

Str1kez commented on Makefile:
```diff
@@ -0,0 +1,52 @@
+	@docker-compose rm --force watchy-db watchy-redis
+endif
+	@docker-compose rm --force watchy-nginx watchy-api
+	@find . -name \*.pyc -delete
```
откуда они появляются?

Str1kez commented on Makefile:
```diff
@@ -0,0 +1,52 @@
+restart: stop start
+
+clean: stop
+ifndef NO_DB_RESET
```
двойное отрицание плохо смотрится `ifdef DB_RESET` лучше наверное

Str1kez commented on Makefile:
```diff
@@ -0,0 +1,52 @@
+	@docker-compose run --rm watchy-api python ./manage.py collectstatic
+endif
+
+cli:
+	@docker-compose run --rm watchy-api bash
```
Странно, для чего это? Что мешает зайти в рабочий контейнер через `docker-compose exec -ti`?

PhonikTop replied:
А это не праивльно? Мне просто так удобнее, меньше печатать.

Str1kez commented on Makefile:
```diff
@@ -0,0 +1,52 @@
+createsuperuser:
+	@docker-compose run --rm watchy-api python ./manage.py createsuperuser
+
+tail:
+	@docker-compose logs -f
```
Зачем они вообще нужны тебе были?) Вроде ничего сложного написать `docker-compose <action>`

PhonikTop replied:
Ну меньше печатать просто)

Str1kez commented on clocks/rooms/redis_client.py:
```diff
@@ -0,0 +1,154 @@
+            votes[uuid] = {"nickname": nickname, "vote": vote}
+            cache.set(self.votes_key, votes, timeout=self.ttl)
+
+    def get_user(self, user_uuid: str) -> Optional[Dict[str, Optional[str]]]:
```
```suggestion
    def get_user(self, user_uuid: str) -> dict[str, str | None] | None
```

Str1kez replied:
Уже давно не надо использовать `Optional`

Str1kez commented on clocks/rooms/redis_client.py:
```diff
@@ -0,0 +1,154 @@
+
+        user_key = f"user:{uuid}:data"
+
+        cache.set(user_key, {"role": role, "nickname": nickname, "vote": vote}, timeout=self.ttl)
```
лучше в явную структуру (dataclass, namedtuple). Данные же в одном виде всегда

Str1kez commented on clocks/rooms/redis_client.py:
```diff
@@ -0,0 +1,154 @@
+        if self.get_user(uuid):
+            raise ValidationError({"error": "User already exists in the room"})
+
+        user_key = f"user:{uuid}:data"
```
в нескольких методах этот ключ создаешь, в одном месте поменяешь – везде искать будешь?

Str1kez commented on clocks/users/serializers.py:
```diff
@@ -0,0 +1,19 @@
+
+class UserInputSerializer(serializers.Serializer):
+    nickname = serializers.CharField(max_length=25)
+    role = serializers.ChoiceField(choices=[("observer", "observer"), ("voter", "voter")])
```
вынеси в отдельный класс чойсов `UserRoleChoices(TextChoices):`, раз есть определенный набор ролей. Этот класс можно использовать в других методах

Str1kez commented on clocks/rooms/redis_client.py:
```diff
@@ -0,0 +1,154 @@
+        self.votes_key = f"{self.room_key}:votes"
+        self.ttl = ttl
+
+    def add_user(self, uuid: str, role: str, nickname: str, vote: Optional[str] = None) -> None:
```
```suggestion
    def add_user(self, uuid: str, role: UserRoleChoices, nickname: str, vote: str | None = None) -> None:
```

Str1kez commented on clocks/rooms/redis_client.py:
```diff
@@ -0,0 +1,154 @@
+        """
+        return cache.get(self.votes_key, {})
+
+    def get_votes_dict(self) -> Dict[str, str]:
```
Какой-то бесполезный метод
- Полностью включает твой метод выше (и почему-то его даже не использует)
- Название очень плохое

Str1kez commented on clocks/rooms/redis_client.py:
```diff
@@ -0,0 +1,154 @@
+        user_key = f"user:{user_uuid}:data"
+        return cache.get(user_key)
+
+    def remove_user(self, user_uuid: str) -> None:
```
Транзакция не нужна? Один удалится, а в списке останется...

Str1kez commented on clocks/rooms/redis_client.py:
```diff
@@ -0,0 +1,154 @@
+from rest_framework.exceptions import ValidationError
+
+
+class RoomCacheManager:
```
Общие проблемы:
- Нет атомарности операций (иногда в методах она необходима)
- Использование устаревшей типизации (Dict, Optional, etc)
- Докстринги – лишнее, чтобы было понятнее лучше использовать [type aliases](https://docs.python.org/3/library/typing.html#type-aliases)
- Это уже не слой работы с данными, а скорее сервис, надо назвать подходяще и в другой модуль закинуть

Str1kez commented on clocks/ws/actions.py:
```diff
@@ -0,0 +1,47 @@
+
+class SubmitVoteAction(BaseAction):
+    async def get_queryset(self):
+        return Meeting.objects.filter(room=self.consumer.lookup_id, active=True).select_related()
```
передать в `select_related` забыл? И `async` не нужен

Str1kez commented on clocks/users/views.py:
```diff
@@ -0,0 +1,36 @@
+    """
+    authentication_classes = [SessionIDAuthentication]
+    serializer_class = UserInputSerializer
+    queryset = Room.objects.all()
```
наверное только активные нужны

Str1kez commented on clocks/users/serializers.py:
```diff
@@ -0,0 +1,19 @@
+    role = serializers.ChoiceField(choices=[("observer", "observer"), ("voter", "voter")])
+
+    def validate(self, data):
+        room = self.context.get("room")
+        if not room:
+            raise ValidationError({"error": "Room instance is required in context"})
```
тебе в контекст всегда будет приходить, или 404 отдаваться

Str1kez commented on clocks/users/serializers.py:
```diff
@@ -0,0 +1,19 @@
+        if not room:
+            raise ValidationError({"error": "Room instance is required in context"})
+
+        current_meeting = Meeting.objects.filter(room_id=room.id, active=True).first()
+        if not current_meeting:
```
```suggestion
        if not Meeting.objects.filter(room_id=room.id, active=True).exists():
```

Str1kez commented on clocks/rooms/redis_client.py:
```diff
@@ -0,0 +1,154 @@
+        :param nickname: Никнейм пользователя.
+        :param vote: Голос пользователя (если есть).
+        """
+        if self.get_user(uuid):
```
Лишний метод, он должен проверять просто наличие в кэше

Str1kez commented on clocks/rooms/views.py:
```diff
@@ -0,0 +1,72 @@
+
+    def get(self, request, pk):
+        room_cache = RoomCacheManager(pk)
+        participants = room_cache.get_room_users()
```
А почему решил вынести участников в кэш, а не хранить как раньше?

PhonikTop replied:
Подумал, что так будет удобнее и структура будет лучше.

Str1kez commented on clocks/ws/actions.py:
```diff
@@ -0,0 +1,47 @@
+        room_cache = RoomCacheManager(self.consumer.lookup_id)
+
+        participants, votes = await sync_to_async(
+            lambda: (room_cache.get_users_by_role("voter"), room_cache.get_votes())
```
почему нормально не разбить на две строки с вызовами `sync_to_async`?

PhonikTop committed:
feat: Реализация аутентификации JWT, сервиса игровых сессий пользователей и кэширования комнат

PhonikTop committed:
feat: Добавление сервиса отправки сообщений

PhonikTop committed:
refactor(makefile): Invert DB reset logic and remove pyc cleanup

PhonikTop committed:
chore(docker): Update postgres image to latest

PhonikTop committed:
feat(room cache service): Add method to get user key

PhonikTop committed:
refactor(room cache service): Use dataclasses for user and vote data, improve data structure

PhonikTop committed:
refactor(websocket actions): Split async cache calls

PhonikTop committed:
refactor(API): Add an active room filter in queryset for the JoinRoom endpoint

PhonikTop committed:
refactor(Websocket SubmitVoteAction): Remove select_related and make get_queryset synchronous

PhonikTop committed:
refactor(UserInputSerializer): Use TextChoices for roles

PhonikTop committed:
refactor(UserInputSerializer): Use exists() instead of first() in validate method

PhonikTop committed:
refactor(RoomCacheService): Improve storage of vote data in cache using dataclasses

PhonikTop committed:
chore(settings): configure CORS headers and middleware

PhonikTop committed:
chore(routing): disable deprecated WebSocket test page

PhonikTop committed:
fix(websocket actions): correct response key from 'status' to 'type'

PhonikTop committed:
fix(websocket): broadcast user join events to all room participants

PhonikTop committed:
feat(api): add endpoint to fetch user details

PhonikTop committed:
feat(frontend): initialize Vue.js project with basic setup

PhonikTop committed:
chore(project): update .gitignore for Vue

PhonikTop committed:
chore(project): remove redundant .gitignore from frontend

PhonikTop committed:
chore(infra): exclude frontend-specific files in .gitignore

PhonikTop committed:
feat(api): setup axios with global configuration and interceptors

PhonikTop committed:
Merge pull request #6 from PhonikTop/feature/init-vue-frontend

chore(frontend): initialize Vue.js project with basic setup

PhonikTop committed:
feat(auth): add login page with basic room joining logic

PhonikTop committed:
feat(room): add room page basics

PhonikTop committed:
chore(router): configure login and room routes

PhonikTop committed:
chore: remove unused HomePage component

PhonikTop committed:
chore(lint): enable setup compiler macros

PhonikTop committed:
Merge pull request #7 from PhonikTop/feature/base-login-room-pages

feat(room): add login and room basic pages

PhonikTop committed:
feat(voting): add VotingForm component with validation

PhonikTop committed:
feat(voting): add ResultsOverlay skeleton

PhonikTop committed:
feat(room): add ObserversList and VotersList with data binding

PhonikTop committed:
feat(room): integrate voting components with state management

PhonikTop committed:
Merge pull request #8

feat(room): add voting interface and participant lists

PhonikTop committed:
fix(websocket): resolve sync_to_async misuse in vote submission action

**Error Description:**
- A `TypeError: sync_to_async can only be applied to sync functions` occurred because `sync_to_async` was incorrectly wrapping the *result* of method calls (e.g., `room_cache.get_users_by_role("voter")`) instead of the methods themselves.
- Affected code: Methods in `SubmitVoteAction.perform_action` interacting with `RoomCacheService`.

PhonikTop committed:
fix(websocket): remove async from get_queryset to resolve improper await usage

PhonikTop committed:
revert(RoomCacheService): remove dataclasses for User and Vote storage

Temporarily revert to dictionary-based storage due to implementation issues.

Dataclasses caused critical malfunctions in the cache service.
Old dictionary structure restored to ensure service stability.

This is a temporary solution until dataclasses can be properly redesigned and reimplemented.

PhonikTop committed:
feat: add useRoom composable

PhonikTop committed:
feat: implement room list fetching via useRoom

PhonikTop committed:
feat: implement room participants fetching via useRoom

PhonikTop committed:
fix: update API baseURL to valid value

PhonikTop committed:
feat: add useUser composable

PhonikTop committed:
feat: implement room join API integration

PhonikTop committed:
feat: implement user verification for room entry

PhonikTop committed:
feat: add useMeeting composable

PhonikTop committed:
feat: implement meeting start

PhonikTop committed:
fix: fix allVoted not working due to missing .value on participants

PhonikTop committed:
feat: add event emitters for meeting control buttons

PhonikTop committed:
feat: implement saving user id to the variable currentUserId

PhonikTop committed:
feat: implement meeting handler methods for ResultOverlay

PhonikTop committed:
fix(serializer): remove redundant validate method

PhonikTop committed:
feat(user-session): include user ID in session response data

PhonikTop committed:
chore: apply code style corrections

PhonikTop committed:
refactor: reorganize imports and variable declarations

PhonikTop committed:
feat: add centralized navigation method

PhonikTop committed:
perf: optimize voting state logic

PhonikTop committed:
feat: add token validation

PhonikTop committed:
refactor: use userError instead of error

PhonikTop committed:
fix: terminate execution after redirect for unauthorized access

PhonikTop committed:
refactor: move fetchParticipants to onBeforeMount

PhonikTop committed:
Merge pull request #9 from PhonikTop/feature/rest-api-integration

feat(frontend): add REST API integration to frontend

PhonikTop committed:
fix(websocket): handle missing 'results' type in response

PhonikTop committed:
refactor(websocket): restructure messages with adding user uuid

PhonikTop committed:
refactor(API): adding active meeting id in answer of RoomDetailView

PhonikTop committed:
fix(API): include missing user_uuid in notify_user_joined call

PhonikTop committed:
fix: fix HTTP methods and request parameters for meeting endpoints

PhonikTop committed:
fix: redirectToLogin in handleEndMeeting is now correct

PhonikTop committed:
feat: add useWebSocket composable

PhonikTop committed:
feat: voting integration using WebSocket

PhonikTop committed:
Merge pull request #10 from PhonikTop/feature/websocket-integration

feat: Realtime voting via WebSocket

PhonikTop committed:
feat: implemented saving and using currentMeeting

PhonikTop committed:
feat: hide voter and observer lists during results phase

PhonikTop committed:
feat: display average_score and resultsVotes in ResultsOverlay component

PhonikTop committed:
feat: implemented restarting and ending a meeting via API

PhonikTop committed:
Merge pull request #11 from PhonikTop/feature/meeting-management

feat: Meeting management and display of voting results

PhonikTop committed:
feat: added ConnectionStatus component to display connection status

PhonikTop committed:
feat: added ChangeTaskForm component for editing meeting task name

PhonikTop committed:
Merge pull request #12 from PhonikTop/feature/voting-ui-widgets

feat(frontend): add missing components for RoomPage

PhonikTop committed:
fix(meeting): prevent division by zero on empty votes

PhonikTop committed:
feat(ws): use jwt-token instead of user_id to send votes

PhonikTop committed:
feat: add online tracking service for participants

PhonikTop committed:
feat: implement channel-user mapping service

PhonikTop committed:
Merge pull request #14 from PhonikTop/feature/participant-presence-and-channels-mapping

PhonikTop committed:
feat: extend room message service with cache integration and new events

PhonikTop committed:
feat: integrate room websocket notifications for meeting events

PhonikTop committed:
refactor: remove deprecated args from RoomMessageService instantiation and calls

PhonikTop committed:
feat: complete RoomConsumer overhaul with auth, services and event handling

PhonikTop committed:
chore: изменить TTL jwt-токена пользователей на 5 часов.

PhonikTop committed:
feat: очищать данные о присутствии участников в комнате после полного окончания встречи

PhonikTop committed:
feat: внедрить room_cache_service для управления данными комнат и пользователей

PhonikTop committed:
feat: внедрить RoomMessageService для отправки сообщений через DjangoChannelMessageSender

PhonikTop committed:
feat: реализовать завершение голосования с сохранением данных комнаты

PhonikTop committed:
feat: отправлять уведомления о перезапуске через RoomMessageService

PhonikTop committed:
feat: очищать голоса в комнате при перезапуске голосования

PhonikTop committed:
feat: использовать множества для данных списка участников комнаты

PhonikTop committed:
feat: реализовать удаление голоса пользователя

PhonikTop committed:
feat: автоматически удалять голос пользователя при его удалении из кэша

PhonikTop committed:
feat: реализовать переход пользователя в другую комнату

PhonikTop committed:
feat: отправлять уведомления об изменении онлайн-статуса пользователя

PhonikTop committed:
feat: реализовать websocket action для изменения статуса голосвания

PhonikTop committed:
feat: отправлять уведомления об перезапуске голосования

PhonikTop committed:
fix: исправить объявление IsAdminUser

PhonikTop committed:
chore: вынести ALLOWED_HOSTS в переменные окружения

PhonikTop committed:
chore: вынести CSRF_TRUSTED_ORIGINS в переменные окружения

PhonikTop committed:
fix: исправить проверку завершённости голосования

PhonikTop committed:
fix: обеспечить асинхронное выполнение синхронных операций

PhonikTop committed:
fix: исправить некорректный URL

PhonikTop committed:
fix: внедрить передачу X-CSRFToken в заголовках API-запросов

PhonikTop committed:
feat: добавить лимит попыток reconnect и экспоненциальную задержку

PhonikTop committed:
refactor: адаптировать VoterList под новую структуру данных голосов (массив ID)

PhonikTop committed:
feat: реализовать синхронизацию состояния голосования через WebSocket

PhonikTop committed:
chore: добавить clocks.env в .gitignore

PhonikTop committed:
infra: добавить Docker-образ для Vue-приложения и Nginx

PhonikTop committed:
infra: упростить инфраструктуру до одного docker-compose файла

PhonikTop committed:
refactor: упростить Dockerfile для backend

PhonikTop committed:
chore: улучшить Makefile с автоматизацией и обновленным синтаксисом Docker

PhonikTop committed:
chore(git): stop tracking clocks.env

PhonikTop committed:
docs: add README.md

PhonikTop committed:
docs: исправить пример значения DEBUG в README

PhonikTop committed:
docs: исправить README.md

PhonikTop committed:
docs: указать название .env файла

PhonikTop committed:
feat: добавлена поддержка drf-spectacular для генерации схем OpenAPI

PhonikTop committed:
docs: добавлена документация для Meetings

PhonikTop committed:
docs: добавлена документация для Rooms

PhonikTop committed:
docs: добавлена документация для Users

PhonikTop committed:
chore: удален устаревший openapi.yaml

PhonikTop committed:
Merge pull request #15 from PhonikTop/feat/api-docs

feat: добавлена OpenAPI документация

PhonikTop committed:
feat: удалена ручка /v1/<int:pk>/history/

PhonikTop committed:
fix: добавлен drf-spectacular в requirements.txt

PhonikTop committed:
chore: удалена конфигурация неиспользуемого black

PhonikTop committed:
chore: обновлена конфигурация ruff

PhonikTop committed:
build: разделено окружение на dev и prod с поддержкой в Makefile

PhonikTop committed:
build: разделено окружение на dev и prod

PhonikTop committed:
chore: удален устаревший TestPage

PhonikTop committed:
style: автоформатирование кода через ruff

PhonikTop committed:
style: автоформатирование кода через ruff

PhonikTop committed:
fix: ошибка "Apps aren't loaded yet" — инициализация Django

PhonikTop committed:
chore: добавлены переменные prod окружения в .qitignore

PhonikTop committed:
fix: исправлены ошибки связанные с параметром

PhonikTop committed:
refactor: переименован метод очистки оффлайн пользователей

PhonikTop committed:
feat: очищать оффлайн пользователей комнаты при перезапуске голосования

PhonikTop committed:
refactor: удален vote в данных пользователя

PhonikTop committed:
refactor: рефакторинг проекта Vue.js

PhonikTop committed:
feat: передавать данные пользователя в сообщениях о смене его статуса онлайн

PhonikTop committed:
refactor: рефакторинг проекта Vue.js

PhonikTop committed:
fix: исправлено отображение taskName

PhonikTop committed:
chore: прекращено отслеживание файла

PhonikTop committed:
chore: обновлены переменные dev окружения

PhonikTop committed:
build: добавлена поддержка переменных окружения в docker-compose для разных сред

PhonikTop committed:
feat: добавлена прокрутка при переполнении списка голосующих

PhonikTop committed:
fix: исправлена отправка сообщений в WebSocket

PhonikTop committed:
feat: добавлены toasts

PhonikTop committed:
style: изменён цвет контента в теме

PhonikTop committed:
fix: исправлена отправка сообщений

PhonikTop committed:
refactor: перенесен room_online_tracker.py в rooms

PhonikTop committed:
refactor: добавлен фильтр на активное голосование

PhonikTop committed:
fix: добавлен 82 порт в CSRF_TRUSTED_ORIGINS для dev-среды

PhonikTop committed:
docs: обновлен README.md

PhonikTop committed:
refactor: удалены избыточные интерфейсы

PhonikTop committed:
refactor: заменен устаревший синтаксис

PhonikTop committed:
refactor: заменен устаревший синтаксис

PhonikTop committed:
style: автоформатирование кода через ruff

PhonikTop committed:
fix: добавлен отсутствующий импорт

PhonikTop committed:
refactor: удален vote из данных пользователя

PhonikTop committed:
feat: добавлен TypedDict для данных пользователя в кеше

PhonikTop committed:
chore: обновлены исключения для ruff

PhonikTop committed:
fix: исправлена ошибка "token invalid"

PhonikTop committed:
style: автоформатирование кода через ruff

PhonikTop committed:
fix: исправлена отправка response из ws actions

PhonikTop committed:
feat: добавлен UserRole для ролей пользователей

PhonikTop committed:
docs: обновлена OpenApi документация

PhonikTop committed:
docs: обновлен README.md

PhonikTop committed:
style: относительные импорты заменены на абсолютные

PhonikTop committed:
refactor: settings.get_list заменен на settings.core.get_env_param_list

PhonikTop committed:
style: автоформатирование кода через ruff

PhonikTop committed:
chore: удален избыточный Makefile
